Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.Statement.cs

```diff
@@ -180,14 +180,61 @@ protected void TrStmt(Statement stmt, ConcreteSyntaxTree wr, ConcreteSyntaxTree
             break;
           }
         case ExpectStmt expectStmt: {
-            var s = expectStmt;
             // TODO there's potential here to use target-language specific features such as exceptions
             // to make it more target-language idiomatic and improve performance
-            ConcreteSyntaxTree bodyWriter = EmitIf(out var guardWriter, false, wr);
-            var negated = new UnaryOpExpr(s.Tok, UnaryOpExpr.Opcode.Not, s.Expr);
-            negated.Type = Type.Bool;
-            EmitExpr(negated, false, guardWriter, wStmts);
-            EmitHalt(s.Tok, s.Message, bodyWriter);
+            if (
+              Options.Get(CommonOptionBag.Verbose) &&
+              expectStmt.Expr is BinaryExpr { Op: BinaryExpr.Opcode.Eq, ResolvedOp: var resolvedOp, E0: var e0, E1: var e1 }) {
+              // If it finds "expect a == b", it will rewrite the code to
+              // var _e0 = a;
+              // var _e1 = b;
+              // if _e0 != _e1 {
+              //   print "Left:"
+              //   print _e0;
+              //   print "Right:"
+              //   print _e1;
+              //   <Halt statement>
+              // }
+              var e0Name = ProtectedFreshId("_e0");
+              var e1Name = ProtectedFreshId("_e1");
+              var e0Var = new LocalVariable(new RangeToken(Token.NoToken, Token.NoToken), e0Name, e0.Type, false);
+              var e1Var = new LocalVariable(new RangeToken(Token.NoToken, Token.NoToken), e1Name, e0.Type, false);
+              DeclareLocalVar(IdName(e0Var), null, e0.Tok, e0, false, wr);
+              DeclareLocalVar(IdName(e1Var), null, e1.tok, e1, false, wr);
+              var e0Ident = new IdentifierExpr(e0.tok, e0Name) {
+                Type = e0.Type,
+                Var = e0Var
+              };
+              var e1Ident = new IdentifierExpr(e1.tok, e0Name) {
+                Type = e1.Type,
+                Var = e1Var
+              };
+
+              ConcreteSyntaxTree bodyWriter = EmitIf(out var guardWriter, false, wr);
+              var negated = new UnaryOpExpr(expectStmt.Tok, UnaryOpExpr.Opcode.Not,
+                new BinaryExpr(expectStmt.Expr.tok, BinaryExpr.Opcode.Eq,
+                  e0Ident,
+                  e1Ident) {
+                  ResolvedOp = resolvedOp,
+                  Type = Type.Bool
+                }) {
+                Type = Type.Bool
+              };
+              EmitExpr(negated, false, guardWriter, wStmts);
+              EmitPrintStmt(bodyWriter, new StringLiteralExpr(e0.tok, "Left:\\n", false) { Type = Type.String() });
```

Comment:
These don't print correctly on the other backends for some reason: I'm getting output like `Test: ['L', 'e', 'f', 't', ':', '\n']1['R', 'i', 'g', 'h', 't', ':', '\n']2`

It might be because `Type.String()` is unresolved?

I think this also suggests Rust is not checking `type.IsStringType` as the other backends are. But you'll have to correct that in order to get tests like `UnicodeString.dfy` to pass eventually. :)

Reply:
Which `type` are you referring to?

Reply:
https://github.com/dafny-lang/dafny/blob/master/Source/DafnyCore/Backends/CSharp/CsharpCodeGenerator.cs#L1995

Reply:
Oh I see ! I changed the type of the string literal to `new SeqType(new CharType())` instead of `Type.String()` so that it will trigger correct printing in other backends. This will work in Rust as well.
I'm still leaving the Rust-specific test.